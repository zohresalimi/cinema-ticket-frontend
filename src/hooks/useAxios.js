import { useEffect, useReducer, useMemo, useCallback } from "react";
import axios from "axios";
import qs from "qs";

const baseUrl = process.env.REACT_APP_BASE_URL;
const actions = {
  REQUEST_START: "REQUEST_START",
  REQUEST_END: "REQUEST_END",
};

async function fetchData({ path, ...config }, dispatch) {
  try {
    dispatch({ type: actions.REQUEST_START });
    const result = await axios({
      ...config,
      url: `${baseUrl}${path}`,
      paramsSerializer: (params) => {
        return qs.stringify(params);
      },
    });
    dispatch({ type: actions.REQUEST_END, payload: result.data });
  } catch (err) {
    dispatch({
      type: actions.REQUEST_END,
      payload: err,
      error: true,
    });
  }
}

const useAxios = (path, config) => {
  function reducer(state, action) {
    switch (action.type) {
      case actions.REQUEST_START:
        return {
          ...state,
          response: null,
          loading: true,
          error: null,
        };
      case actions.REQUEST_END:
        return {
          ...state,
          loading: false,
          ...(action.error ? {} : { data: action.payload.data }),
          [action.error ? "error" : "response"]: action.payload,
        };
      default:
        return state;
    }
  }

  function createInitialState({ manual }) {
    return {
      loading: !manual,
      error: null,
      response: null,
    };
  }

  const options = useMemo(
    () => ({
      data: {},
      method: "get",
      manual: false,
      path,
      ...config,
    }),
    [config, path]
  );
  const [state, dispatch] = useReducer(reducer, createInitialState(options));

  useEffect(() => {
    if (!options.manual) {
      fetchData(options, dispatch);
    }
  }, [options, dispatch, path]);

  const refetch = useCallback(
    (payload) => {
      return fetchData({ ...options, ...payload }, dispatch);
    },
    [options, dispatch]
  );

  return [state, refetch];
};

export default useAxios;
