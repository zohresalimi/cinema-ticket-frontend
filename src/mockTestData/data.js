import React from "react";
import { Router } from "@reach/router";
import AppContext from "../store/context";
import Theme from "../Styles/Theme";

export const getTestStore = () => {
  return {
    dispatch: () => {},
    state: {
      user: {
        name: "",
        email: "",
      },
      movies: {
        children: [
          {
            _id: "5fa71823adc463397a50c501",
            actors: ["Holland Roden", "Ronen Rubinstein", "Keegan Allen"],
            age: "7",
            coverImage:
              "https://catalog.cinema-api.com/cf/images/ncg-images/68bfeef55ada4e7ab0606db2af0f40ad.jpg?width=240&version=EA2E80EE161E5DA161C6E26487054C9F&format=webp",
            description:
              "Poppy and Kvist embark on an adventure that will take them far beyond",
            director: "aaron woodley",
            duration: "1:31",
            genre: ["family", "children"],
            images: [],
            largeImage:
              "https://catalog.cinema-api.com/cf/images/ncg-images/79276cead82341d2a004f8e8f99a88e9.jpg?width=1920&version=08F0C83177D59FB94C4C809F63F67E2A&format=webp",
            name: "trolls 2: world tour",
            originalTitle: "swedish",
            premiere: "2020-10-03T15:18:52.722Z",
            rooms: ["5fa653d0a7f1201999dd1ac0"],
            trailerUrl: "https://youtu.be/P9A6hOg9QQ0",
          },
        ],
        premiered: [
          {
            _id: "5fa71823adc463397a50c502",
            actors: ["Holland Roden", "Ronen Rubinstein", "Keegan Allen"],
            age: "7",
            coverImage:
              "https://catalog.cinema-api.com/cf/images/ncg-images/68bfeef55ada4e7ab0606db2af0f40ad.jpg?width=240&version=EA2E80EE161E5DA161C6E26487054C9F&format=webp",
            description:
              "Poppy and Kvist embark on an adventure that will take them far beyond",
            director: "aaron woodley",
            duration: "1:31",
            genre: ["family", "children"],
            images: [],
            largeImage:
              "https://catalog.cinema-api.com/cf/images/ncg-images/79276cead82341d2a004f8e8f99a88e9.jpg?width=1920&version=08F0C83177D59FB94C4C809F63F67E2A&format=webp",
            name: "follow me",
            originalTitle: "swedish",
            premiere: "2020-10-05T15:18:52.722Z",
            rooms: ["5fa653c0643c06586bba7d78"],
            trailerUrl: "https://youtu.be/P9A6hOg9QQ0",
          },
        ],
        upcoming: [
          {
            _id: "5fa71823adc463397a50c503",
            actors: ["Holland Roden", "Ronen Rubinstein", "Keegan Allen"],
            age: "7",
            coverImage:
              "https://catalog.cinema-api.com/cf/images/ncg-images/68bfeef55ada4e7ab0606db2af0f40ad.jpg?width=240&version=EA2E80EE161E5DA161C6E26487054C9F&format=webp",
            description:
              "Poppy and Kvist embark on an adventure that will take them far beyond",
            director: "aaron woodley",
            duration: "1:31",
            genre: ["family", "children"],
            images: [],
            largeImage:
              "https://catalog.cinema-api.com/cf/images/ncg-images/79276cead82341d2a004f8e8f99a88e9.jpg?width=1920&version=08F0C83177D59FB94C4C809F63F67E2A&format=webp",
            name: "honest thief",
            originalTitle: "swedish",
            premiere: "2021-01-03T15:18:52.722Z",
            rooms: ["5fa653c0643c06586bba7d78"],
            trailerUrl: "https://youtu.be/P9A6hOg9QQ0",
          },
        ],
        allMovies: {
          "5fa71823adc463397a50c501": "children",
          "5fa71823adc463397a50c502": "premiered",
          "5fa71823adc463397a50c503": "upcoming",
        },
      },
      cinemas: [
        {
          _id: "5fa64fe234fbbf9945f3a3ce",
          name: "filmstaden kista",
          purchaseEndTime: "21:30",
          purchaseStartTime: "8:00",
          rooms: ["5fa653c0643c06586bba7d78", "5fa653d0a7f1201999dd1ac0"],
        },
      ],
      ticket: {
        movie: {},
        status: false,
        quantity: 0,
        price: 0,
        seatNumbers: new Map(),
        user: {},
        showing: {},
      },
      showings: [
        {
          _id: "5fbe9bd31f58462b5c759b5c",
          capacity: 4,
          cinema: "5fa64fe234fbbf9945f3a3ce",
          endTime: null,
          movie: {
            _id: "5fa71823adc463397a50c502",
            actors: ["Holland Roden", "Ronen Rubinstein", "Keegan Allen"],
            age: "7",
            description:
              "Poppy and Kvist embark on an adventure that will take them far beyond",
            director: "aaron woodley",
            duration: "1:31",
            genre: ["family", "children"],
            images: [],
            name: "follow me",
            originalTitle: "swedish",
            premiere: "2020-10-05T15:18:52.722Z",
            rooms: ["5fa653c0643c06586bba7d78"],
            trailerUrl: "https://youtu.be/P9A6hOg9QQ0",
          },
          price: 100,
          room: { _id: "5fa653c0643c06586bba7d78" },

          seats: [
            [
              { seatNumber: 1, taken: "available" },
              { seatNumber: 1, taken: "available" },
            ],
            [
              { seatNumber: 1, taken: "available" },
              { seatNumber: 1, taken: "available" },
            ],
          ],
          startTime: "2020-11-13T16:00:00.000Z",
        },
      ],
    },
  };
};

export const WithProvider = (props) => {
  const { state, dispatch } = getTestStore();
  return (
    <Theme>
      <AppContext.Provider
        value={{ state: props.defaultValue || state, dispatch }}
      >
        <Router>{props.children}</Router>
      </AppContext.Provider>
    </Theme>
  );
};
