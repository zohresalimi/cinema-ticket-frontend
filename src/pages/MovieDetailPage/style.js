import styled from "styled-components";

const Wrapper = styled.div`
  font-family: ${(props) => props.theme.fontFamily.sf};
  display: flex;
  align-items: center;
  width: 100%;
  flex-direction: column;
  margin-top: 40px;
  font-size: ${(props) => props.theme.fontSizes.content};
  color: ${(props) => props.theme.colors.textColor};
  line-height: 25px;
  p {
    color: ${(props) => props.theme.colors.white};
    font-size: ${(props) => props.theme.fontSizes.content};
    margin: 5px 0;
    &.description {
      margin-bottom: 30px;
    }
  }
  .title {
    font-size: ${(props) => props.theme.fontSizes.small};
    color: ${(props) => props.theme.colors.gray};
    &:after {
      content: ":";
    }
  }
  i {
    margin: 0 10px;
  }
  .displayInline {
    p {
      display: inline;
      margin-right: 5px;
      &:after {
        content: ",";
      }
    }
  }
  h2.buyTicket {
    margin: 100px 0 45px 0;
  }

  .cinema-name {
    font-size: 1.45rem;
    color: #9c9c9c;
  }

  .showings-list {
    list-style: none;
    display: flex;
    align-items: center;
    padding: 0;
    justify-content: space-between;
    li {
      border-bottom: 1px solid #929292;
      width: 100%;
      display: flex;
      padding: 10px 20px;
      &:hover {
        background-color: rgb(53 53 53 / 80%);
      }
      a {
        text-decoration: none;
        display: flex;
        width: 100%;
        justify-content: space-between;
        .time-slot {
          flex: 1;
        }
        .room-name {
        }
        .movie-subtitle {
          color: #888;
        }
      }
    }
  }
  .buy-ticket p {
    color: ${(props) => props.theme.colors.red};
  }
`;

export { Wrapper as default };
