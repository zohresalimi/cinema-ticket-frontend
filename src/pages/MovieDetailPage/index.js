import React, { useState, useContext, useRef, useEffect } from "react";
import { Link } from "@reach/router";
import Moment from "react-moment";
import { useTranslation } from "react-i18next";
import TopBar from "../../components/TopBar";
import MovieDetail from "../../components/MovieDetail";

import useAxios from "../../hooks/useAxios";
import AppContext from "../../store/context";
import {
  SET_SELECTED_MOVIE_REDUCER,
  SET_SHOWINGS_MOVIE_REDUCER,
  SET_CINEMAS_REDUCER,
} from "../../constants";
import { Container, Row } from "../../Styles/StyleComponents";
import Wrapper from "./style";

function MovieDetailPage(props) {
  const { state, dispatch } = useContext(AppContext);
  const { t, i18n } = useTranslation();
  const { category, movieId } = props;
  const { movies } = state;
  const [currentMovie] = useState(() => {
    return movies[category].find((movie) => movie._id === movieId);
  });

  const dataRef = useRef({
    params: { rooms: currentMovie.rooms.join(",") },
  });

  const [{ response: cinemas }] = useAxios(
    `/api/v1/cinemas/by-room-ids`,
    dataRef.current
  );

  const [{ response: showings }] = useAxios(
    `/api/v1/showings/by-movie-id/${currentMovie._id}`
  );

  const showingByCinemaId = (cinemaId) => {
    return showings.data.filter((showing) => showing.room.cinema === cinemaId);
  };

  useEffect(() => {
    if (cinemas) {
      dispatch({ type: SET_CINEMAS_REDUCER, data: cinemas.data });
    }
  }, [cinemas, dispatch]);

  useEffect(() => {
    if (currentMovie) {
      dispatch({ type: SET_SELECTED_MOVIE_REDUCER, data: currentMovie });
    }
  }, [currentMovie, dispatch]);

  useEffect(() => {
    if (showings) {
      dispatch({ type: SET_SHOWINGS_MOVIE_REDUCER, data: showings.data });
    }
  }, [showings, dispatch]);

  return (
    <div>
      <TopBar />
      <MovieDetail movie={currentMovie} />
      <div className="black-bg">
        <Container>
          <Row>
            <Wrapper>
              {currentMovie && (
                <div>
                  <p className="description">{t(currentMovie.description)}</p>
                  <p className="title">{t("director")}</p>
                  <p>{t(currentMovie.director)}</p>
                  <p className="title">{t("premiere")}</p>
                  <p>
                    <Moment
                      locale={i18n.language}
                      date={currentMovie.premiere}
                      format="ddd D MMM"
                    />
                  </p>
                  <p className="title">{t("actor")}</p>
                  <div className="displayInline">
                    {currentMovie.actors.map((actor) => (
                      <>
                        <p key={actor}>{actor}</p>
                      </>
                    ))}
                  </div>

                  <p className="title">{t("original language")}</p>
                  <p>{currentMovie.originalTitle}</p>
                  <h2 className="buyTicket">{t("book tickets")}</h2>
                  <div>
                    {cinemas &&
                      cinemas.data.map((item) => {
                        return (
                          <div key={item._id}>
                            <h2 className="cinema-name">{item.name}</h2>
                            <ul className="showings-list">
                              {showings &&
                                showingByCinemaId(item._id).map((el) =>
                                  el.capacity > 0 ? (
                                    <li key={el._id}>
                                      <Link
                                        to={`../../../booking/${el._id}`}
                                        state={{ category }}
                                      >
                                        <div className="showing-info">
                                          <p className="time-slot">
                                            <Moment
                                              date={el.startTime}
                                              format="hh:mm"
                                            />
                                          </p>
                                          <p className="room-name">
                                            {t(el.room.name)}
                                          </p>
                                          <p className="movie-subtitle">
                                            {t("lang")}
                                            {":"}
                                            {t(currentMovie.originalTitle)}
                                          </p>
                                        </div>
                                        <div className="buy-ticket">
                                          <p>{t("buy ticket")}</p>
                                        </div>
                                      </Link>
                                    </li>
                                  ) : (
                                    <li key={el._id}>
                                      <p className="room-name">
                                        <Moment
                                          date={el.startTime}
                                          format="hh:mm"
                                        />
                                      </p>
                                      <p className="room-name">
                                        {t(el.room.name)}
                                      </p>
                                      <p className="buy-ticket">
                                        {t(currentMovie.originalTitle)}
                                      </p>
                                    </li>
                                  )
                                )}
                            </ul>
                          </div>
                        );
                      })}
                  </div>
                </div>
              )}
            </Wrapper>
          </Row>
        </Container>
      </div>
    </div>
  );
}

export default MovieDetailPage;
