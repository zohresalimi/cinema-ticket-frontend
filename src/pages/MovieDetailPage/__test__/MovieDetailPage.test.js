import * as React from "react";
import { when } from "jest-when";
import { render, act, waitFor, fireEvent } from "@testing-library/react";
import * as hooks from "../../../hooks/useAxios";
import "@testing-library/jest-dom/extend-expect";
import { getTestStore, WithProvider } from "../../../mockTestData/data";
import MovieDetailPage from "../";

jest.mock("react-i18next", () => ({
  useTranslation: () => ({
    t: (k) => k,
    i18n: {
      language: "sv",
      changeLanguage: () => new Promise(() => {}),
    },
  }),
}));

async function renderWrapper(setDefaultState, props) {
  let component;
  const defaultValue = setDefaultState
    ? {
        user: {},
        movies: {
          upcoming: [],
          children: [],
          premiered: [],
          allMovies: {},
        },
        cinemas: [],
        showings: [],
        ticket: {},
      }
    : null;

  act(() => {
    component = render(
      <WithProvider mockDispatch={() => {}} defaultValue={defaultValue}>
        <MovieDetailPage {...props} path="/" />
      </WithProvider>
    );
  });

  return component;
}

jest.mock("../../../components/TopBar", () => () => (
  <div>TopBar Component</div>
));
jest.mock("../../../components/MovieDetail", () => (props) => (
  <div>
    <button onClick={props.onBookingClick} data-testid="button-scroll">
      scroll
    </button>
    MovieDetail Component
  </div>
));

describe("MovieDetail page Component Testing", () => {
  afterAll(() => {
    jest.clearAllMocks();
  });

  beforeEach(() => {
    jest.clearAllMocks();
  });

  test("take snapshot", async () => {
    const { container } = await renderWrapper(true);
    expect(container.firstChild).toMatchSnapshot();
  });

  test("should call scrollToBookingSection", async () => {
    const mockScrollIntoView = jest.fn();
    const mockCreateRef = jest.fn(() => ({
      current: { scrollIntoView: mockScrollIntoView },
    }));

    jest.spyOn(React, "createRef").mockImplementation(mockCreateRef);

    const { getByTestId } = await renderWrapper();
    fireEvent.click(getByTestId("button-scroll"));
    expect(mockScrollIntoView).toHaveBeenCalled();
  });

  test("should read current movie from state", async () => {
    const testState = getTestStore();
    const movieId = testState.movies.premiered[0]._id;

    const { container } = await renderWrapper(false, {
      category: "premiered",
      movieId,
    });
    expect(container.firstChild).toMatchSnapshot();
  });

  test("should show loading and call useAxios 3 times", async () => {
    const testState = getTestStore();
    const movieId = testState.movies.premiered[0]._id;
    const mockUseAxios = jest.fn();
    when(mockUseAxios)
      .calledWith("/api/v1/cinemas/by-room-ids", { manual: true })
      .mockReturnValue([
        {
          response: null,
        },
        jest.fn(),
      ]);
    when(mockUseAxios)
      .calledWith(`/api/v1/showings/by-movie-id/${movieId}`)
      .mockReturnValue([
        {
          response: null,
          loading: true,
        },
      ]);
    when(mockUseAxios)
      .calledWith("", { manual: true })
      .mockReturnValue([
        {
          response: null,
        },
        jest.fn(),
      ]);

    jest.spyOn(hooks, "default").mockImplementation(mockUseAxios);

    const { container } = await renderWrapper(false, {
      category: "premiered",
      movieId,
    });
    await waitFor(() => expect(mockUseAxios).toHaveBeenCalledTimes(3));
    expect(container.firstChild).toMatchSnapshot();
    jest.clearAllMocks();
  });

  test("set current movie by calling useAxios", async () => {
    const mockUseAxios = jest.fn();
    when(mockUseAxios)
      .calledWith("", {
        manual: true,
      })
      .mockReturnValue([
        {
          response: {},
          loading: true,
        },
      ]);
    when(mockUseAxios)
      .calledWith("/api/v1/cinemas/by-room-ids", { manual: true })
      .mockReturnValue([{}, jest.fn()]);
    when(mockUseAxios)
      .calledWith(`/api/v1/showings/by-movie-id/1`)
      .mockReturnValue([{}]);
    jest.spyOn(hooks, "default").mockImplementation(mockUseAxios);

    const mockSetState = jest.fn();
    const mockUseState = jest.fn(() => [false, mockSetState]);
    jest.spyOn(React, "useState").mockImplementation(mockUseState);
    await renderWrapper(true, {
      movieId: 1,
    });
    expect(mockSetState).toHaveBeenCalled();
    jest.clearAllMocks();
  });

  test("should set cinemas data to state", async () => {
    const mockDispatch = jest.fn();
    const mockUseContext = jest.fn(() => ({
      state: {},
      dispatch: mockDispatch,
    }));
    jest.spyOn(React, "useContext").mockImplementation(mockUseContext);

    const mockUseAxios = jest.fn();
    when(mockUseAxios)
      .calledWith("", {
        manual: true,
      })
      .mockReturnValue([
        {
          response: {},
          loading: true,
        },
      ]);
    when(mockUseAxios)
      .calledWith("/api/v1/cinemas/by-room-ids", { manual: true })
      .mockReturnValue([{ response: {} }, jest.fn()]);
    when(mockUseAxios)
      .calledWith(`/api/v1/showings/by-movie-id/1`)
      .mockReturnValue([{ response: {} }]);
    jest.spyOn(hooks, "default").mockImplementation(mockUseAxios);

    const mockSetState = jest.fn();
    const mockUseState = jest.fn(() => [false, mockSetState]);
    jest.spyOn(React, "useState").mockImplementation(mockUseState);
    await renderWrapper(true, {
      movieId: 1,
    });
    expect(mockDispatch).toHaveBeenCalledTimes(2);
    jest.clearAllMocks();
  });
});
