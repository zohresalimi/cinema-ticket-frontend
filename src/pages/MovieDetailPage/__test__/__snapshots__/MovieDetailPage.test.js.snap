// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MovieDetail page Component Testing should read current movie from state 1`] = `
<div
  style="outline: none;"
  tabindex="-1"
>
  <div>
    <div>
      TopBar Component
    </div>
    <div>
      <button
        data-testid="button-scroll"
      >
        scroll
      </button>
      MovieDetail Component
    </div>
    <div
      class="black-bg"
    >
      <div
        class="style__Wrapper-cmlatl-0 bbndKs"
      >
        <div
          class="StyleComponents__Container-sc-11erdxt-0 fFgwWO"
        >
          <div
            class="StyleComponents__Row-sc-11erdxt-1 jhoSVF"
          >
            <div
              class="full-width"
            >
              
              <h2
                class="buyTicket"
              >
                no showings found
              </h2>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`MovieDetail page Component Testing should show loading and call useAxios 3 times 1`] = `
<div
  style="outline: none;"
  tabindex="-1"
>
  <div>
    <div>
      TopBar Component
    </div>
    <div>
      <button
        data-testid="button-scroll"
      >
        scroll
      </button>
      MovieDetail Component
    </div>
    <div
      class="black-bg"
    >
      <div
        class="style__Wrapper-cmlatl-0 bbndKs"
      >
        <div
          class="StyleComponents__Container-sc-11erdxt-0 fFgwWO"
        >
          <div
            class="StyleComponents__Row-sc-11erdxt-1 jhoSVF"
          >
            <div
              class="full-width"
            >
              <h2
                class="buyTicket"
              >
                Loading...
              </h2>
              <h2
                class="buyTicket"
              >
                no showings found
              </h2>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`MovieDetail page Component Testing take snapshot 1`] = `
<div
  style="outline: none;"
  tabindex="-1"
>
  <div>
    <div>
      TopBar Component
    </div>
    <div>
      <button
        data-testid="button-scroll"
      >
        scroll
      </button>
      MovieDetail Component
    </div>
    <div
      class="black-bg"
    >
      <div
        class="style__Wrapper-cmlatl-0 bbndKs"
      >
        <div
          class="StyleComponents__Container-sc-11erdxt-0 fFgwWO"
        >
          <div
            class="StyleComponents__Row-sc-11erdxt-1 jhoSVF"
          />
        </div>
      </div>
    </div>
  </div>
</div>
`;
